src/Avalonia.Base/Collections/Arena.cs

```diff
@@ -0,0 +1,110 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace Avalonia.Collections;
+
+/// <summary>
+/// Reduces memory allocations by handing out <see cref="Span{T}"/> views of larger array chunks.
+/// </summary>
+/// <remarks>
+/// Intended for scenarios in which many short-lived collections are required simultaneously.
+/// </remarks>
+internal class Arena<T>
```

Comment:
Am I missing something or is this just a worse version of an array pool? 
It doesn't really support reentrancy since it tries to use a single array per thread, allocates unpooled memory  and uses GCHandle for a reason that escapes me

Reply:
It also doesn't seem to clear the array once it's freed, which can lead to memory leaks

Reply:
1) I still don't see where the array gets cleared. The "current" array is GC-rooted, so it will cause a temporary memory leak (which can and will GC-root huge trees of objects in user apps)
2) Previously we didn't allocate at all. Why can't we use an ArrayPool? Why do we need to be constantly allocating new arrays?